source/common/tcp_proxy/upstream.cc

```diff
@@ -119,6 +119,7 @@ void HttpUpstream::resetEncoder(Network::ConnectionEvent event, bool inform_down
   if (inform_downstream) {
     upstream_callbacks_.onEvent(event);
```

Comment:
can we end up having an event callback "from the upstream connection" before there is a connection established.  That's a bit weird.  I think it might be cleaner if inform_downstream is true, that if we have a deferer we do pool failure, and if not we do the onEvent. WDYT?